Sources/NIOSSL/SSLCertificateExtensions.swift

```diff
@@ -0,0 +1,197 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@_implementationOnly import CNIOBoringSSL
+@_implementationOnly import CNIOBoringSSLShims
+
+public struct _SSLCertificateExtensions {
+    private enum Storage {
+        final class Deallocator {
+            let reference: OpaquePointer!
+            
+            init(takeOwnershipOf reference: OpaquePointer!) {
+                self.reference = reference
+            }
+            
+            deinit {
+                CNIOBoringSSL_sk_X509_EXTENSION_free(self.reference)
+            }
+        }
+        
+        case owned(Deallocator)
+        case borrowed(reference: OpaquePointer!, owner: AnyObject)
+        
+        init(takeOwnershipOf reference: OpaquePointer!) {
+            self = .owned(.init(takeOwnershipOf: reference))
+        }
+        
+        init(borrowing reference: OpaquePointer!, owner: AnyObject) {
+            self = .borrowed(reference: reference, owner: owner)
+        }
+        
+        /// The owner of the memory to which the reference points
+        var owner: AnyObject {
+            switch self {
+            case .owned(let deallocator):
+                return deallocator
+            case .borrowed(_, let owner):
+                return owner
+            }
+        }
+        
+        /// All operations accessing `reference` need to be implemented while guaranteeing that we still have a reference to the memory owner.
+        /// Otherwise `reference` could already be freed. This would result in undefined behaviour as we access a dangling pointer.
+        /// This method guarantees that `reference` is valid during execution of `body`.
+        internal func withReference<Result>(
+            _ body: (OpaquePointer?) throws -> Result
+        ) rethrows -> Result {
+            try withExtendedLifetime(self) {
+                switch self {
+                case .owned(let deallocator):
+                    return try body(deallocator.reference)
+                case .borrowed(let reference, _):
+                    return try body(reference)
+                }
+            }
+        }
+    }
+    
+    @usableFromInline internal let stackSize: Int
+    private let storage: Storage
+    
+    internal init(takeOwnershipOf reference: OpaquePointer!) {
+        self.storage = .init(takeOwnershipOf: reference)
+        self.stackSize = CNIOBoringSSL_sk_X509_EXTENSION_num(reference)
+    }
+    
+    internal init(borrowing reference: OpaquePointer!, owner: AnyObject) {
+        self.storage = .init(borrowing: reference, owner: owner)
+        self.stackSize = CNIOBoringSSL_sk_X509_EXTENSION_num(reference)
+    }
+    
+    internal init(copyOf reference: OpaquePointer!) {
+        self.init(takeOwnershipOf: CNIOBoringSSL_sk_X509_EXTENSION_dup(reference))
+    }
+    
+    internal init() {
+        self.init(takeOwnershipOf: nil)
```

Comment:
We need to support the empty case [because CNIOBoringSSL_X509_get0_extensions returns nil if no extensions are present](https://www.openssl.org/docs/man1.1.1/man3/X509_get0_extensions.html). The other option we have would be to make the `NIOSSLCertificate._extensions` optional which I don't like.

Reply:
Do we need an empty `init` like this at `internal` scope?